urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/RegExp.java

```diff
@@ -0,0 +1,866 @@
+			a = BasicOperations.union(list);
+			a.minimize();
+			break;
+		case REGEXP_CONCATENATION:
+			list = new ArrayList<Automaton>();
+			findLeaves(exp1, Kind.REGEXP_CONCATENATION, list, automata, automaton_provider, minimize);
+			findLeaves(exp2, Kind.REGEXP_CONCATENATION, list, automata, automaton_provider, minimize);
+			a = BasicOperations.concatenate(list);
+			a.minimize();
+			break;
+		case REGEXP_INTERSECTION:
+			a = exp1.toAutomaton(automata, automaton_provider, minimize).intersection(exp2.toAutomaton(automata, automaton_provider, minimize));
+			a.minimize();
+			break;
+		case REGEXP_OPTIONAL:
+			a = exp1.toAutomaton(automata, automaton_provider, minimize).optional();
+			a.minimize();
+			break;
+		case REGEXP_REPEAT:
+			a = exp1.toAutomaton(automata, automaton_provider, minimize).repeat();
+			a.minimize();
+			break;
+		case REGEXP_REPEAT_MIN:
+			a = exp1.toAutomaton(automata, automaton_provider, minimize).repeat(min);
+			a.minimize();
+			break;
+		case REGEXP_REPEAT_MINMAX:
+			a = exp1.toAutomaton(automata, automaton_provider, minimize).repeat(min, max);
+			a.minimize();
+			break;
+		case REGEXP_COMPLEMENT:
+			a = exp1.toAutomaton(automata, automaton_provider, minimize).complement();
+			a.minimize();
+			break;
+		case REGEXP_CHAR:
+			a = BasicAutomata.makeChar(c);
+			break;
+		case REGEXP_CHAR_RANGE:
+			a = BasicAutomata.makeCharRange(from, to);
+			break;
+		case REGEXP_ANYCHAR:
+			a = BasicAutomata.makeAnyChar();
+			break;
+		case REGEXP_EMPTY:
+			a = BasicAutomata.makeEmpty();
+			break;
+		case REGEXP_STRING:
+			a = BasicAutomata.makeString(s);
+			break;
+		case REGEXP_ANYSTRING:
+			a = BasicAutomata.makeAnyString();
+			break;
+		case REGEXP_AUTOMATON:
+			Automaton aa = null;
+			if (automata != null)
+				aa = automata.get(s);
+			if (aa == null && automaton_provider != null)
+				try {
+					aa = automaton_provider.getAutomaton(s);
+				} catch (IOException e) {
+					throw new IllegalArgumentException(e);
+				}
+			if (aa == null)
+				throw new IllegalArgumentException("'" + s + "' not found");
+			a = aa.clone(); // always clone here (ignore allow_mutate)
+			break;
+		case REGEXP_INTERVAL:
+			a = BasicAutomata.makeInterval(min, max, digits);
+			break;
+		}
+		return a;
+	}
+
+	private void findLeaves(RegExp exp, Kind kind, List<Automaton> list, Map<String, Automaton> automata, 
+			AutomatonProvider automaton_provider,
+			boolean minimize) {
+		if (exp.kind == kind) {
+			findLeaves(exp.exp1, kind, list, automata, automaton_provider, minimize);
+			findLeaves(exp.exp2, kind, list, automata, automaton_provider, minimize);
+		} else
+			list.add(exp.toAutomaton(automata, automaton_provider, minimize));
+	}
+
+	/** 
+	 * Constructs string from parsed regular expression. 
+	 */
+	@Override
+	public String toString() {
+		return toStringBuilder(new StringBuilder()).toString();
+	}
+
+	StringBuilder toStringBuilder(StringBuilder b) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 50 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 98% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.